L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+most discussed._ Server credentials will mostly just mirror the channel
+credentials; they suffer from most of the same problems but at a smaller
+scale.
+
+## Background
+
+Prior to gRPC v0.12.0, C core had a "unified" `Credential` type that was
+simultaneously a call and channel credential. This was the design proposed for
+all languages. However, the Java implementation could not determine how such an
+API could be implemented for Java with its multiple transports. The designer of
+the API had too little bandwidth (it was a _very_ busy time), so Java (and Go)
+developed its own API. Java used interceptors (which no other language had) and
+credential-specific configuration on the ManagedChannelBuilder.
+
+In gRPC v0.12.0, C core swapped to a “split” CallCredential and
+ChannelCredential. ChannelCredentials could contain CallCredentials. At this
+point it was not a priority for Java to investigate the new API; Java was happy
+they didn’t use the previous API, such that they didn’t have an API migration.
+
+gRPC Java 0.15.0 added CallCredentials to replace the interceptor-based call
+credentials. At this point Java had looked into the full “split” credential
+design, but decided it was too hard to implement at that moment. Java worked to
+stabilize the `ManagedChannelBuilder.usePlaintext()` and
+`useTransportSecurity()` APIs.
+
+Go has split call credentials (`credentials.PerRPCCredentials`) and channel
+credentials (`credentials.TransportCredentials`), but with a separate type to
+combine them (`credentials.Bundle`). The split credentials significantly
+predate `Bundle`.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+`ChannelCredentials` will be introduced, which can be provided when
+constructing a `ManagedChannelBuilder':
+
+```java
+// instead of:
+channel = ManagedChannelBuilder.forTarget("example.com")
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
+    .build();
+```
+
+`newChannelBuilder()` will iterate through the `ManagedChannelProviders`
+trying to create a builder for each in turn. The first provider that succeeds
+will be used. If no provider is able to handle the credentials, then
+`newChannelBuilder()` will throw. `newChannelBuilderForAddress()` will be a
+convenience function that creates the target string and calls
+`newChannelBuilder()`, mirroring its current behavior.
+
+The new ChannelCredentials API cannot be mixed with the pre-existing
+`useTransportSecurity()` and `usePlaintext()` builder APIs. The old-style
```

Comment:
Consider at some point deprecating them when this proposal is merged.

Reply:
Yes, plan to do so, eventually. But it will be slow-moving. There are a _lot_ of usages of this API, so I want to make sure there is some natural adoption and see that things are working well before strongly encouraging a migration.